### PR TITLE
Algos & Strategies params + Benders phases

### LINKED ISSUE #107: Remove functions setalgorithm and getalgorithm from StrategyRecord.
Do we still need the structure StrategyRecord?

### LINKED ISSUE #113: Update/clean the code regarding the roles of algorithms/strategies
- methods `run!(::AbstractAlgorithm)!` should only modify the formulations and should return an object of type `AbstractAlgorithmRecord`
- methods `apply!(::AbstractStrategy)` are allowed to modify the node according to the record returned by the algorithms.
- Should methods `prepare!(::AbstractAlgorithm)` create the object `AbstractAlgorithmData` and return it  to the function `apply!(::AbstractAlgorithm)`? `apply!(::AbstractAlgorithm)` would then call the methods `run!(::AbstractAlgorithm)` passing only the algorithm type and its data.

### PR DESCRIPTION
**Pull request by guimarqu**
fix #113 
fix #107

### CODE REVIEW

File: src/algorithms/algorithm.jl [hunk 1]
```diff
@@ -1,60 +1,53 @@
...
+    set_algorithm_result!(node, algo, record)
     return record
-end
-
-"""
-    StartNode
-
-Fake algorithm that indicates the start of the node treatment.
-"""
-struct StartNode <: AbstractAlgorithm end
```
Comment (vitornesello) [inline src/algorithms/algorithm.jl]: why did we have this and why we don't need it anymore?

File: src/algorithms/benders.jl [hunk 2]
```diff
@@ -1,9 +1,14 @@
...
+    feasibility_tol::Float64 = 1e-5
+    optimality_tol::Float64 = 1e-5
+    max_nb_iterations::Int = 100
 end
-BendersCutGeneration() = BendersCutGeneration(false)
 
-mutable struct BendersCutGenerationData
+#BendersCutGeneration() = BendersCutGeneration(false)
+
+mutable struct BendersCutGenTmpRecord
```
Comment (vitornesello) [inline src/algorithms/benders.jl]: can we think of a better name for this?
maybe : 
`RuntimeRecord` or
`AuxiliaryRecord`

File: src/algorithms/benders.jl [hunk 3]
```diff
@@ -13,36 +18,43 @@ mutable struct BendersCutGenerationData
     #slack_cost_increase_applied::Bool
 end
 
-function allphasesapplied(algdata::BendersCutGenerationData)
+function allphasesapplied(algdata::BendersCutGenTmpRecord)
```
Comment (vitornesello) [inline src/algorithms/benders.jl]: this function names is very hard to read. Can we use `all_phases_applied` ?

File: src/algorithms/benders.jl [hunk 4]
```diff
@@ -52,40 +64,40 @@ function update_bendersep_slackvar_cost_for_ph1!(spform::Formulation)
...
-            #setcurcost!(spform, var, 0.0)
+            setcurcost!(spform, var, 0.0)
         end
+        # TODO if previous phase is  a pure phase 2, reset current ub
     end
     return
 end
 
-update_bendersep_slackvar_cost_for_ph2!(spform::Formulation) = return
+function update_bendersep_slackvar_cost_for_ph2!(spform::Formulation)
```
Comment (vitornesello) [inline src/algorithms/benders.jl]: what is `bendersep`

File: src/algorithms/benders.jl [hunk 5]
```diff
@@ -95,39 +107,69 @@ function update_bendersep_problem!(
...
+        elseif phase_to_apply == PurePhase2
+            update_bendersep_slackvar_cost_for_ph2!(spform)
+        end
+        algdata.spform_phase_applied[spform_uid] = true
+    end
+    return false
+end
+
+function reset_bendersep_phase!(algdata::BendersCutGenTmpRecord, reform::Reformulation)
+
```
Comment (vitornesello) [inline src/algorithms/benders.jl]: blank line

File: src/algorithms/benders.jl [hunk 6]
```diff
@@ -95,39 +107,69 @@ function update_bendersep_problem!(
...
 
-function update_bendersep_target!(sp_form::Formulation)
+function update_bendersep_target!(spform::Formulation)
     # println("bendersep target will only be needed after automating convexity constraints")
 end
 
 
-function insert_cuts_in_master!(algdata, master_form::Formulation,
-                                sp_form::Formulation,
+function insert_cuts_in_master!(algdata, masterform::Formulation,
```
Comment (vitornesello) [inline src/algorithms/benders.jl]: not following styleguide

File: src/algorithms/benders.jl [hunk 7]
```diff
@@ -175,44 +217,54 @@ function insert_cuts_in_master!(algdata, master_form::Formulation,
     return nb_of_gen_cuts
 end
 
-function compute_bendersep_pb_contrib(
-    algdata::BendersCutGenerationData, spform::Formulation, spsol::OptimizationResult{S}
+function compute_bendsp_lagrangian_bound_contrib(
```
Comment (vitornesello) [inline src/algorithms/benders.jl]: I think `bendersep` and `bendsp` refer to "benders subproblem". Can we use `benders_sp` everywhere?

File: src/algorithms/benders.jl [hunk 8]
```diff
@@ -175,44 +217,54 @@ function insert_cuts_in_master!(algdata, master_form::Formulation,
...
     contrib = getvalue(dualsol)
     return contrib
 end
 
 function solve_sp_to_gencut!(
-    algdata::BendersCutGenerationData, masterform::Formulation, 
+    algo::BendersCutGeneration, algdata::BendersCutGenTmpRecord, masterform::Formulation, 
     spform::Formulation, master_primal_sol::PrimalSolution{S},
-    master_dual_sol::DualSolution{S}
+    master_dual_sol::DualSolution{S}, upto_phase::FormulationPhase
```
Comment (vitornesello) [inline src/algorithms/benders.jl]: `upto_phase` -> `up_to_phase` ?

File: src/algorithms/preprocessing.jl [hunk 9]
```diff
@@ -73,29 +73,29 @@ function change_sp_bounds!(alg_data::PreprocessData)
...
 
     for (col, col_val) in alg_data.local_partial_sol
-        sp_form = getsp(alg_data, col)
-        sp_form_id = getuid(sp_form)
-        if alg_data.cur_sp_bounds[sp_form_id][1] > 0
-            alg_data.cur_sp_bounds[sp_form_id] = (
-                max(alg.cur_sp_bounds[sp_form_id][1] - col_val, 0),
-                alg.cur_sp_bounds[sp_form_id][2]
+        spform = getsp(alg_data, col)
+        spform_id = getuid(spform)
```
Comment (vitornesello) [inline src/algorithms/preprocessing.jl]: why `spform_id`?

File: src/algorithms/reformulationsolver.jl [hunk 10]
```diff
@@ -20,7 +20,7 @@ was_fully_explored(t::SearchTree) = t.fully_explored
 """
     ReformulationSolver
 
-Branch-and-bound algorithm to optimize a `Reformulation`.
+Branch-and-bound algorithm to optimize a `Reformulationjuli.
```
Comment (vitornesello) [inline src/algorithms/reformulationsolver.jl]: `Reformulationjuli` ?

File: src/formulation.jl [hunk 11]
```diff
@@ -195,13 +195,13 @@ function setprimaldwspsol!(master_form::Formulation,
                          is_explicit::Bool = true,
                          moi_index::MoiVarIndex = MoiVarIndex()) where {S<:AbstractObjSense}
 
-    #master_form = sp_form.parent_formulation
-    mastcol_id = generatevarid(master_form)
+    #masterform = spform.parent_formulation
+    mastcol_id = generatevarid(masterform)
```
Comment (vitornesello) [inline src/formulation.jl]: `mastcol_id`? I don't understand the naming standards

File: src/formulation.jl [hunk 12]
```diff
@@ -427,15 +427,21 @@ function remove_from_optimizer!(ids::Set{Id{T}}, f::Formulation) where {
 end
 
 function resetsolvalue(form::Formulation, sol::PrimalSolution{S}) where {S<:AbstractObjSense}
-    val = sum(getperenecost(getvar(form, var_id)) * value for (var_id, value) in sol)
-    setvalue!(sol, val)
-    return val
+    if length(sol) > 0
+        val = sum(getperenecost(getvar(form, var_id)) * value for (var_id, value) in sol)
+        setvalue!(sol, val)
```
Comment (vitornesello) [inline src/formulation.jl]: Shouldn't you set the value even if `length(sol) == 0`?

File: src/formulation.jl [hunk 13]
```diff
@@ -427,15 +427,21 @@ function remove_from_optimizer!(ids::Set{Id{T}}, f::Formulation) where {
...
+        return val
+    end
+    return 0.0
 end
 
 function resetsolvalue(form::Formulation, sol::DualSolution{S}) where {S<:AbstractObjSense}
```
Comment (vitornesello) [inline src/formulation.jl]: can't we have `resetsolvalue` for an `AbstractSolution`?

Comment (vitornesello) [inline src/formulation.jl]: Should this function really return `val`?

File: src/incumbents.jl [hunk 14]
```diff
@@ -102,16 +102,42 @@ the current one according to the objective sense.
 """
 function set_ip_dual_bound!(inc::Incumbents{S},
```
Comment (vitornesello) [inline src/incumbents.jl]: does not respect styleguide

File: src/strategies/strategy.jl [hunk 15]
```diff
@@ -65,17 +56,7 @@ A GlobalStrategy encapsulates all three strategies necessary to define Coluna's
...
 """
 struct GlobalStrategy <: AbstractStrategy
-    conquer::Type{<:AbstractConquerStrategy}
-    divide::Type{<:AbstractDivideStrategy}
-    tree_search::Type{<:AbstractTreeSearchStrategy}
+    conquer::AbstractConquerStrategy
+    divide::AbstractDivideStrategy
+    tree_search::AbstractTreeSearchStrategy
 end
-
```
Comment (vitornesello) [inline src/strategies/strategy.jl]: Why this was deleted?

File: src/algorithms/algorithm.jl (hunk 1 above)
Comment (guimarqu) [reply src/algorithms/algorithm.jl]: It was to initialize the current algorithm fied in the strategy record. We deleted the strategy record.

File: src/algorithms/benders.jl (hunk 2 above)
Comment (guimarqu) [reply src/algorithms/benders.jl]: Maybe we should keep Data in that case ?

File: src/algorithms/benders.jl (hunk 4 above)
Comment (guimarqu) [reply src/algorithms/benders.jl]: benders separation (benders_sep)

File: src/algorithms/benders.jl (hunk 3 above)
Comment (guimarqu) [reply src/algorithms/benders.jl]: I removed the function, it was dead code.

File: src/formulation.jl (hunk 11 above)
Comment (guimarqu) [reply src/formulation.jl]: `mast_col_id`

File: src/formulation.jl (hunk 13 above)
Comment (guimarqu) [reply src/formulation.jl]: Primal uses cost, Dual uses rhs to compute the value.

Comment (guimarqu) [reply src/formulation.jl]: I think we should because it is a value computed in the function.

File: src/strategies/strategy.jl (hunk 15 above)
Comment (guimarqu) [reply src/strategies/strategy.jl]: Because it was commented

File: src/algorithms/benders.jl (hunk 2 above)
Comment (vitornesello) [reply src/algorithms/benders.jl]: to see qith francois, i think he doesnt like data

File: src/strategies/strategy.jl (hunk 15 above)
Comment (vitornesello) [reply src/strategies/strategy.jl]: why was it commented? i thought we used it

File: src/algorithms/benders.jl (hunk 2 above)
Comment (guimarqu) [reply src/algorithms/benders.jl]: @FD-V

**Pull request merged**